youtube_dl/extractor/youtube.py

```diff
@@ -215,10 +228,21 @@ def warn(message):
 
                 tfa_results = req(
                     self._TFA_URL.format(tl), tfa_req,
-                    'Submitting TFA code', 'Unable to submit TFA code')
+                    'Submitting TFA code', 'Unable to submit TFA code, trying another way')
 
                 if tfa_results is False:
-                    return False
+                    tfa_req = [
+                        user_hash, None, 2, None,
+                        [
+                            6, None, None, None, None,
+                            [tfa_code, True]
+                        ]]
+                    tfa_results = req(
+                        self._TFA_URL.format(tl), tfa_req,
+                        'Submitting TFA code', 'Unable to submit TFA code')
```

Comment:
You should use exactly the method youtube expects you to, not try all methods in a row.

Reply:
YouTube expects different things from different people. For you it wants the first payload (`[None, tfa_code, True, 2]`, for others it wants the second (`[tfa_code, True]`). As far as I can tell there is no way to know which one it wants. The only way to get it to work for everyone is to try both methods.

Reply:
It's obviously possible since youtube ask you to use some particular method when logging in via web and not tries all the methods available. Trying all methods in a row only approaches risk of ban due to number of tries.

Reply:
I can't find anything in the packets that indicate which of the payloads it wants. Maybe if I had packets from someone who youtube wants `[None, tfa_code, True, 2]` from I could see the difference.

Reply:
This is the response text right after I select TOTP as the TFA method, so the next request would be submitting the TOTP code:
```JSON
[
  [
    [
      "gf.siscr",
      [
        null,null,
        "TWO_STEP_VERIFICATION",
        2,null,
        "INITIALIZED",
        null,null,
        6,7,1,1,1,null,null,
        "PRIVATE (my name)",
        "PRIVATE (email)",
        "https://lh3.googleusercontent.com/MAYBE/PRIVATE/IDK/mo/photo.jpg",
        null,null,null,null,
        [],
        {
          "1006":["IPHONE"],
          "5006":[]
        }
      ]
    ],
    [
      "gf.ttu",
      0,
      "PRIVATE (TL)"
    ],
    [
      "e",
      3,null,null,
      447
    ]
  ]
]
```
Nothing here particularly stands out to me as indicating which response format it wants.

Reply:
You can select preferred method when logging in via web. Then you'll have the payloads for this method.
I tried to figure out this several times either along with adding support for other TFA methods but each time I've been hit by TFA code limit so I gave up on this.

Reply:
Oh, this isn't about different TFA methods. This is completely about TOTP.

[This block](https://github.com/ytdl-org/youtube-dl/pull/18521/files/9f7a57dfc34b0801abf27e9ec4863245c094a034#diff-bd8242a0122c5207531954b67e6e51f0R210) ensures that TOTP is selected no matter what. The block we are commenting on is about submitting the TOTP code to youtube, which it wants in different payloads for different people for some reason.

Reply:
Specifically, the request to `https://accounts.google.com/_/signin/challenge?hl=en&TL=...`, which is when the TOTP code gets submitted to youtube.

For me the postData of that request has a property `f.req` with a value of `["user_hash",null,2,null,[6,null,null,null,null,["tfa_code",false]]]`. Based on your comment saying this didn't work your postData's `f.req` needs to look like `["user_hash",null,2,null,[9,null,null,null,null,null,null,null,[null,"tfa_code",false,2]]]`. I can't tell how to find out which is needed, which is why I have it try both.

Reply:
The issue appears to be distinguishing between codes sent over SMS (`[None, tfa_code, True, 2]`), and codes generated from an authenticator app (Google Authenticator, Authy, etc). which is sent using `[tfa_code, True]`.

The list of available challenges is returned (default first) from `challenge_results`, with a corresponding ID (from the above code appears to be available as `select_tfa_results[8]`) and some extra information for each challenge. Rather than switching challenges after failure, you can switch only if the default challenge is not `6` or `9`, and similarly determine which payload to use.

I'd very, very much recommend looking at [this code](https://github.com/omarroth/invidious/blob/6215259/src/invidious.cr#L1004-L1084) for reference, which is heavily based on youtube-dl's implementation.

Reply:
@omarroth Ahh, that makes a lot of sense. What I was doing was forcing the Google Auth (TOTP) challenge, since I thought that was the only one being supported by youtube-dl. Instead I should check which challenge it is and submit based on that, maybe switching to SMS by default if something else is selected?

Doing that might also make it possible to support other challenges, and maybe include a way to pick your challenge.

Reply:
> Instead I should check which challenge it is and submit based on that, maybe switching to SMS by default if something else is selected?

That would be my suggestion. There's also some additional challenges for login verification (if attempting to use an unrecognized device) that would easily be supported, although I expect that's out of scope for this PR.